docs/src/userguide/extension_types.rst

```diff
@@ -355,9 +355,10 @@ you need to explicitly allow :keyword:`None` values for them.
 
 .. note::
 
-    ``not None`` and ``typing.Optional`` can only be used in Python functions (defined with
-    :keyword:`def` and without ``@cython.cfunc`` decorator) and not C functions
-    (defined with :keyword:`cdef` or decorated using ``@cython.cfunc``).  If
+    ``not None`` clause can only be used in Python functions (defined with
+    :keyword:`def`) and not C functions (defined with :keyword:`cdef`).
+    Type annotations can be used in either Python functions and C functions,
+    but :keyword:`None` value is not checked when used in C functions. If
```

Comment:
Hmm. Was this really wrong? I'm not sure if we allow `Optional[]` in C functions. And if it has an effect if we do.
(BTW, it's "The ... clause", because it's a specific one.)

Reply:
Yes it works:
```python
```

Reply:
> (BTW, it's "The ... clause", because it's a specific one.)

fixed

Reply:
> Not sure what to do here for the time being. We'd better not let users run into this trap.

I agree it is difficult situation.I think we should raise a warning with detailed explanation during compiling sources. @da-woods what do you think?

Reply:
I was thinking. Cython currently does not detect `@exceptval` decorator:
```python
@cython.exceptval(check=True)
@cython.cfunc
def foo_nonopt(bar: Bar) -> cython.int:
# We can now check None value of bar and raise exception when has None value
...
```
In this case if cython detects problematic cdef function:
```python
@cython.cfunc
def foo_nonopt(bar: Bar) -> cython.int:
```
We can print an warning (even maybe error - not sure) during compilation like this:
```
WARNING: C function cannot guarantee non-None value of parameter 'bar'.
You should use Optional[] type to allow None values or add @exceptval decorator
to allow propagate exceptions.
```

Reply:
A warning sounds too annoying here. The code will be fine in too many cases, and users won't have a good way to resolve the issue anyway. Let's at least revert the documentation change, because `Optional` really isn't supported here.

Reply:
> Let's at least revert the documentation change

done.